File: hw3/hw3 - vectors/tetromino.cpp
#include "tetromino.h"

//	constructor
Tetromino::Tetromino(Tetrominos shape)
{
	init(shape);
	shapeName = shape;
}

//	rotate function takes direction as a parameter and rotates the tetromino
//  90 degrees in given direction
void Tetromino::rotate(string dir)
{
	
	if(dir == "right")
	{
		vector<vector<char>> temp(this->getTetro()[0].size(), vector<char>(this->getTetro().size()));

		for (int row = 0; row < this->getTetro().size(); row++)
		{
			for (int col = 0; col < this->getTetro()[row].size(); col++)
			{
				temp[col][row] = this->getTetro()[row][col];
			}
			
		}
		for (int row = 0; row < temp.size(); row++)
		{
			for (int col = 0; col < temp[row].size()/2; col++)
			{
				swap(temp[row][col], temp[row][temp[row].size() - col - 1]);
			}
			
		}

		this->getTetro()= temp;

	}
	if(dir == "left")
	{

		for (int row = 0; row < this->getTetro().size(); row++)
		{
			for (int col = 0; col < this->getTetro()[row].size()/2; col++)
			{
				swap(this->getTetro()[row][col], this->getTetro()[row][this->getTetro()[row].size() - col - 1]);
			}
			
		}


		vector<vector<char>> temp(this->getTetro()[0].size(), vector<char>(this->getTetro().size()));

		for (int row = 0; row < temp.size(); row++)
			for (int col = 0; col < temp[row].size(); col++)
				temp[row][col] = this->getTetro()[col][row];

		this->getTetro()= temp;
		
	}
	
}

//	print function prints the current tetromino on the screen
void Tetromino::print()
{
	for (auto &a : this->getTetro())
	{
		for (auto &b : a)
		{
			cout << b;
		}

		cout << "\n";
	}	
}

//	getter function for the tetromino vector
vector<vector<char>> &Tetromino::getTetro()
{
	return tetro;
}

//	getter function for the tetromino name in enum type
Tetrominos Tetromino::getShapeName()
{
	return shapeName;
}

//	defShape function uses enum class to convert char input to enum type
Tetrominos Tetromino::defShape(const char shape)
{
	switch (shape)
	{
	case 'I':
		return Tetrominos::I;
		break;
	case 'O':
		return Tetrominos::O;
		break;
	case 'T':
		return Tetrominos::T;
		break;
	case 'J':
		return Tetrominos::J;
		break;
	case 'L':
		return Tetrominos::L;
		break;
	case 'S':
		return Tetrominos::S;
		break;
	case 'Z':
		return Tetrominos::Z;
		break;					
	}

	return Tetrominos::I;
}

//	init function is called by constructor. it initializes the current tetromino vector
void Tetromino::init(Tetrominos shape)
{
	switch(shape)
	{
		case Tetrominos::I:			
			tetro = {
				{'I', 'I', 'I', 'I'}
			};	
			break;
		case Tetrominos::O:
			tetro = {
				{'O', 'O'},
				{'O', 'O'}
			};	
			break;
		case Tetrominos::T:
			tetro = {
				{'T', 'T', 'T'},
				{' ', 'T', ' '}
			};	
			break;
		case Tetrominos::J:
			tetro = {
				{' ', 'J'},
				{' ', 'J'},
				{'J', 'J'}
			};		
			break;
		case Tetrominos::L:
			tetro = {
				{'L', ' '},
				{'L', ' '},
				{'L', 'L'}
			};		
			break;
		case Tetrominos::S:
			tetro = {
				{' ', 'S', 'S'},
				{'S', 'S', ' '}
			};		
			break;
		case Tetrominos::Z:
			tetro = {
				{'Z', 'Z', ' '},
				{' ', 'Z', 'Z'}
			};		
			break;
	}
}

File: hw7/Tetromino.java
package tetris;

import java.util.Arrays;
import java.util.Random;

/**
 * 
 * @author emre
 */

public class Tetromino {

	private char[][] tetro;
	private int xPos;
	private int yPos;

	/**
	 * constructor
	 * @param shape tetromino shape as char to be created
	 */
	public Tetromino(char shape) {

		init(shape);
		yPos = 0;
		
	}

	/**
	 * getter for the shape array
	 * @return 2d shape char array
	 */
	public char[][] getTetro() {
		return Arrays.copyOf(tetro, tetro.length);
	}

	/**
	 * returns the position of the tetromino horizontally
	 * @return xPos variable
	 */
	public int getxPos() {
		return xPos;
	}

	/**
	 * sets the position of the tetromino horizontally
	 * @param xPos xPos variable
	 */
	public void setxPos(int xPos) {
		this.xPos = xPos;
	}

	/**
	 * returns the position of the tetromino vertically
	 * @return yPos variable
	 */
	public int getyPos() {
		return yPos;
	}

	/**
	 * sets the position of the tetromino vertically
	 * @param yPos yPos variable
	 */
	public void setyPos(int yPos) {
		this.yPos = yPos;
	}
	
	/**
	 * rotates the tetromino shape to the left or right respect to given parameter
	 * @param dir direction to rotate
	 */
	public void rotate(String dir){
		
		if("right".equals(dir))
		{
			char temp[][] = new char[this.tetro[0].length][this.tetro.length];

			for (int row = 0; row < this.tetro.length; row++)
			{
				for (int col = 0; col < this.tetro[row].length; col++)
				{
					temp[col][row] = this.tetro[row][col];
				}

			}
			for (int row = 0; row < temp.length; row++)
			{
				for (int col = 0; col < temp[row].length/2; col++)
				{
					char dump = temp[row][col];
					temp[row][col] = temp[row][temp[row].length - col - 1];
					temp[row][temp[row].length - col - 1] = dump;
				}

			}

			this.tetro = temp;

		}
		if("left".equals(dir))
		{

			for (int row = 0; row < this.tetro.length; row++)
			{
				for (int col = 0; col < this.tetro[row].length/2; col++)
				{
					char dump = this.tetro[row][col];
					this.tetro[row][col] = this.tetro[row][this.tetro[row].length - col - 1];
					this.tetro[row][this.tetro[row].length - col - 1] = dump;
				}

			}

			char temp[][] = new char[this.getTetro()[0].length][this.getTetro().length];

			for (int row = 0; row < this.getTetro().length; row++)
			{
				for (int col = 0; col < this.getTetro()[row].length; col++)
				{
					temp[col][row] = this.getTetro()[row][col];
				}

			}

			this.tetro = temp;

		}
		
	}
	
	/**
	 * initializes the tetromino array
	 * @param shape tetromino shape
	 */
	public final void init(char shape){
		switch (shape) {
			case 'I' ->
				tetro = new char[][]{
					{'I', 'I', 'I', 'I'}
				};
			case 'O' ->
				tetro = new char[][]{
					{'O', 'O'},
					{'O', 'O'}
				};
			case 'T' ->
				tetro = new char[][]{
					{'T', 'T', 'T'},
					{' ', 'T', ' '}
				};
			case 'J' ->
				tetro = new char[][]{
					{' ', 'J'},
					{' ', 'J'},
					{'J', 'J'}
				};
			case 'L' ->
				tetro = new char[][]{
					{'L', ' '},
					{'L', ' '},
					{'L', 'L'}
				};
			case 'S' ->
				tetro = new char[][]{
					{' ', 'S', 'S'},
					{'S', 'S', ' '}
				};
			case 'Z' ->
				tetro = new char[][]{
					{'Z', 'Z', ' '},
					{' ', 'Z', 'Z'}
				};
			case 'R' ->
				init(randShape());
		}
	}
	
	/**
	 * returns a random shape
	 * @return random shape as char
	 */
	public static char randShape(){
		
		Random random = new Random();
		
		int rand = random.nextInt(7) + 1;
		
		switch (rand)
		{
		case 1 -> {
			return 'I';
			}
		case 2 -> {
			return 'O';
			}
		case 3 -> {
			return 'T';
			}
		case 4 -> {
			return 'J';
			}
		case 5 -> {
			return 'L';
			}
		case 6 -> {
			return 'S';
			}
		case 7 -> {
			return 'Z';
			}						
		}
		return 'Q';		
	}

}
